pandas/core/frame.py

```diff
@@ -1620,20 +1651,55 @@ def to_records(self, index=True, convert_datetime64=None):
 
             count = 0
             index_names = list(self.index.names)
+
             if isinstance(self.index, MultiIndex):
                 for i, n in enumerate(index_names):
                     if n is None:
                         index_names[i] = 'level_%d' % count
                         count += 1
             elif index_names[0] is None:
                 index_names = ['index']
+
             names = (lmap(compat.text_type, index_names) +
                      lmap(compat.text_type, self.columns))
         else:
             arrays = [self[c].get_values() for c in self.columns]
             names = lmap(compat.text_type, self.columns)
+            index_names = []
+
+        index_len = len(index_names)
+        formats = []
+
+        for i, v in enumerate(arrays):
+            index = i
+
+            if index < index_len:
+                dtype_mapping = index_dtypes
+                name = index_names[index]
+            else:
+                index -= index_len
+                dtype_mapping = column_dtypes
+                name = self.columns[index]
+
+            if isinstance(dtype_mapping, dict):
```

Comment:
Don't hardcode dict here. Use `compat.Mapping`.

Reply:
actually use ``is_dict_like``

Reply:
Ah, good catch.  I’ll fix that.

Reply:
One thing though: `is_dict_like` does not guarantee existence of the `__contains__` method, which is what my implementation also relies on.

I could do some try-except logic to make it more generic, or I could make the `is_dict_like` check slightly stricter.

Thoughts?

Reply:
I would try and see if you can make is_dict_like stronger I think its ok here

Reply:
I really hate to use other things because they are completely unmaintanable in the codebase, hey just pick your favorite thing to check (dict, collections.Mapping), oh and you need something else, just reinvent it. 

This is the reason we have the is_* things.

Reply:
Okay, sounds good.  I was also leaning towards expanding the dict-like check but wanted to get some consensus.

I do understand the appeal of the collections checking, but as we already have the inference utilities, we should use them in the interest of maintainability.